### Libraries for Installation

`{# {r} # install.packages("BiocManager") # BiocManager::install("ALDEx2") # BiocManager::install(c("phyloseq", "microbiome", "ComplexHeatmap"), update = FALSE) # install.packages("microViz",repos = c(davidbarnett = "https://david-barnett.r-universe.dev", getOption("repos"))) # install.packages("ggraph") # for taxatree_plots() # install.packages("DT") # for tax_fix_interactive() # install.packages("corncob") # for example datasets and beta binomial models # install.packages("patchwork") # install.packages("ggpubr") # install.packages("ggrepel") # install.packages ("cowplot") # install.packages("rnaturalearth") # install.packages("rnaturalearthdata") # install.packages ("ggspatial") # install.packages("maps") # install.packages("microViz")`

```{r}
#install.packages("lmtest")
```

```{r}
#install.packages("tibble")
# install.packages("grDevices")
```

### Add Installed Libraries to Pipeline

```{r}
library(ALDEx2)
library(stringr)
library(phyloseq)
library(ggplot2)
library(vegan)
library(patchwork)
library(knitr)
library(ggpubr)
library(ggrepel)
library(cowplot)
library(microViz)
library(lubridate, warn.conflicts = FALSE)
library(lmtest)
library(tibble)
library(grDevices)

```

### Load Metadata

```{r, warning=FALSE, error=FALSE, message=FALSE, results='hide', fig.show='hide'}
metadata <- read.csv("metadata_updated.csv", row.names=1)

```

### Loading Metaxa2 Results

```{r, warning=FALSE, error=FALSE, message=FALSE, results='hide', fig.show='hide'}
metaxa_genus <- read.delim("metaxa_genus.txt")

# Create OTU table
OTU_metaxa <- metaxa_genus[,-1]
# Match sample ID order with metadata file
match <- match(rownames(metadata), colnames(OTU_metaxa))
OTU_metaxa <- OTU_metaxa[,match]
all(colnames(OTU_metaxa) == rownames(metadata))
# Create tax table
tax_table_metaxa <- data.frame(str_split_fixed(data.frame(metaxa_genus) [,1], ";", 6))
colnames(tax_table_metaxa) <- c("Domain", "Phylum", "Class", "Order", "Family", "Genus")
# Check if samples are in order
identical(rownames(metadata), colnames(OTU_metaxa))
# Combine into phyloseq object
metaxa_PHY <- phyloseq(otu_table(OTU_metaxa, 
    taxa_are_rows=TRUE), tax_table(as.matrix(tax_table_metaxa)), sample_data(metadata))

# Exclude taxa "Unknown", "Unclassified", "Eukaryota", "Mitochondria", "Archaea", "Chloroplast"
metaxa_PHY <- subset_taxa(metaxa_PHY, Domain == "Bacteria")

# Add SSU counts to metadata
metadata$SSU_counts <- sample_sums(metaxa_PHY)

```

### Insert ARG Counts (Resfinder)

```{r, warning=FALSE, error=FALSE, message=FALSE}
counts_resfinder <-as.matrix(read.table("ARG_genemat.txt", header= T, check.names = F, row.names = 1))

# Tax_table
clusters_tax_table_resfinder <- read.csv("clusters_tax_table.txt", header=FALSE, sep=";")
colnames(clusters_tax_table_resfinder) <- c("Gene",  "Cluster_name", "Class")
# Reorder columns
col_order <- c("Class", "Cluster_name", "Gene")
clusters_tax_table_resfinder <- clusters_tax_table_resfinder[, col_order]

# Reorder to match metadata
match <- match(rownames(metadata), colnames(counts_resfinder))
counts_resfinder <- counts_resfinder[,match]
all(colnames(counts_resfinder) == rownames(metadata))

# Reorder to match tax_table
match <- match(rownames(counts_resfinder), clusters_tax_table_resfinder$Gene)
clusters_tax_table_resfinder <- clusters_tax_table_resfinder[match,]
all(rownames(counts_resfinder) == clusters_tax_table_resfinder$Gene)

# remove rownames
counts_resfinder_phy <- counts_resfinder
rownames(counts_resfinder_phy) <- c(1:3146) #PKY: Why hard-coded?
rownames(clusters_tax_table_resfinder) <- c(1:3146) #PKY: Why hard-coded?

# Phyloseq
resfinder_PHY_counts <- phyloseq(otu_table(counts_resfinder_phy, taxa_are_rows = TRUE), 
  sample_data(metadata), tax_table(as.matrix(clusters_tax_table_resfinder)))

# Exclude biological / technical replicates
resfinder_PHY_counts_uniq <- subset_samples(resfinder_PHY_counts, replicate == "NO")
resfinder_PHY_counts_uniq <- subset_samples(resfinder_PHY_counts_uniq, sample_material == "water")

# Create phyloseq object with only HWW samples
resfinder_PHY_counts_HWW <- subset_samples(resfinder_PHY_counts_uniq, HWW == "YES")

```

### Insert ARG Relative Abundances (Resfinder)

```{r, warning=FALSE, error=FALSE, message=FALSE, results='hide', fig.show='hide'}
counts_resfinder <-as.matrix(read.table("ARG_genemat.txt", header= T, check.names = F, row.names = 1))

# Tax_table
clusters_tax_table_resfinder <- read.csv("clusters_tax_table.txt", header=FALSE, sep=";")
colnames(clusters_tax_table_resfinder) <- c("Gene",  "Cluster_name", "Class")
# Reorder columns
col_order <- c("Class", "Cluster_name", "Gene")
clusters_tax_table_resfinder <- clusters_tax_table_resfinder[, col_order]

# Reorder to match metadata
match <- match(rownames(metadata), colnames(counts_resfinder))
counts_resfinder <- counts_resfinder[,match]
all(colnames(counts_resfinder) == rownames(metadata))

# Reorder to match tax_table
match <- match(rownames(counts_resfinder), clusters_tax_table_resfinder$Gene)
clusters_tax_table_resfinder <- clusters_tax_table_resfinder[match,]
all(rownames(counts_resfinder) == clusters_tax_table_resfinder$Gene)

# Divide by ARG gene lengths
resfinder_lengths <- read.delim("resfinder_lengths.txt", header=FALSE, comment.char="#")
all(rownames(clusters_tax_table_resfinder$Gene) == resfinder_lengths$V1)
resfinder_length_norm <- counts_resfinder/resfinder_lengths[, 2]

# Normalization with Metaxa2 SSU counts
resfinder_length_SSU_norm <- t(t(resfinder_length_norm)/metadata$SSU_counts) * 1540
all(rownames(metadata) == colnames(resfinder_length_SSU_norm))
identical((resfinder_length_norm[3, 5]/metadata$SSU_counts[5]) * 1540, resfinder_length_SSU_norm[3, 5])
all(rownames(resfinder_length_norm) == clusters_tax_table_resfinder$Gene)
# Hide rownames
dim(resfinder_length_SSU_norm)
rownames(resfinder_length_SSU_norm) <- c(1:3146) #PKY: Why hard-coded?

dim(clusters_tax_table_resfinder)
rownames(clusters_tax_table_resfinder) <- c(1:3146) #PKY: Why hard-coded?

# Combine to phyloseq object
resfinder_PHY_rel <- phyloseq(otu_table(resfinder_length_SSU_norm, taxa_are_rows = TRUE),
  sample_data(metadata), tax_table(as.matrix(clusters_tax_table_resfinder)))

# Exclude biological / technical replicates
resfinder_PHY_rel_uniq <- subset_samples(resfinder_PHY_rel, replicate == "NO")
resfinder_PHY_rel_uniq <- subset_samples(resfinder_PHY_rel_uniq, sample_material == "water")

# Create phyloseq object with only HWW samples
resfinder_PHY_rel_HWW <- subset_samples(resfinder_PHY_rel_uniq, HWW == "YES")
```

### Metaphlan Counts and Relative Abundances

```{r, warning=FALSE, error=FALSE, message=FALSE, results='hide', fig.show='hide'}
metaphlan_counts_merged_abundance_table <- read.delim("merged_metaphlan_count_species.txt", header = T)
mp_counts <- metaphlan_counts_merged_abundance_table[ -c(1) ]

# Tax table
## Check if in order
tax_table_metaphlan_counts <- read.csv2("tax_table_count_metaphlan_species.txt", header=F, sep="")

tax_table_metaphlan_counts <- data.frame(str_split_fixed(data.frame(tax_table_metaphlan_counts) [,1], ";", 7))
colnames(tax_table_metaphlan_counts) <- c("Kingdom", "Phylum", "Class", "Order", "Family", "Genus", "Species")

tax_table_metaphlan_counts <- apply(tax_table_metaphlan_counts, 2, function(y) (gsub(".__", "", y)))

# Reorder
match <- match(rownames(metadata), colnames(mp_counts))
mp_counts  <- mp_counts[,match]
all(rownames(metadata) == colnames(mp_counts))

# Combine into phyloseq object
metaphlan_PHY_counts <- phyloseq(otu_table(mp_counts, taxa_are_rows=TRUE), sample_data(metadata), tax_table(as.matrix(tax_table_metaphlan_counts), errorIfNULL = TRUE))

# Leave only bacteria
metaphlan_PHY_counts_bact <- subset_taxa(metaphlan_PHY_counts, Kingdom != "Viruses" & Kingdom != "Eukaryota" & Kingdom != "Archaea")

# Prune to species level
metaphlan_PHY_counts_genus <- tax_glom(metaphlan_PHY_counts_bact, taxrank = "Genus")

# Prune to diff taxa level
metaphlan_PHY_counts_species <- tax_glom(metaphlan_PHY_counts_bact, taxrank = "Species")
metaphlan_PHY_counts_genus <- tax_glom(metaphlan_PHY_counts_bact, taxrank = "Genus")
metaphlan_PHY_counts_family <- tax_glom(metaphlan_PHY_counts_bact, taxrank = "Family")
metaphlan_PHY_counts_order <- tax_glom(metaphlan_PHY_counts_bact, taxrank = "Order")
metaphlan_PHY_counts_class <- tax_glom(metaphlan_PHY_counts_bact, taxrank = "Class")
metaphlan_PHY_counts_phylum <- tax_glom(metaphlan_PHY_counts_bact, taxrank = "Phylum")
```

```{r, warning=FALSE, error=FALSE, message=FALSE, results='hide', fig.show='hide'}
metaphlan_merged_abundance_table <- read.delim("merged_metaphlan_rel_species.txt", header = T)
mp_rel <- metaphlan_merged_abundance_table[ -c(1) ]

# Tax table
## Check if in order
tax_table_metaphlan <- read.csv2("tax_table_metaphlan_rel_species.txt", header=F, sep="")

tax_table_metaphlan <- data.frame(str_split_fixed(data.frame(tax_table_metaphlan) [,1], ";", 7))
colnames(tax_table_metaphlan) <- c("Kingdom", "Phylum", "Class", "Order", "Family", "Genus", "Species")

tax_table_metaphlan <- apply(tax_table_metaphlan, 2, function(y) (gsub(".__", "", y)))

# Reorder
match <- match(rownames(metadata), colnames(mp_rel))
mp_rel  <- mp_rel[,match]
all(rownames(metadata) == colnames(mp_rel))

# Combine into phyloseq object
metaphlan_PHY_rel <- phyloseq(otu_table(mp_rel, taxa_are_rows=TRUE), sample_data(metadata), tax_table(as.matrix(tax_table_metaphlan)))

# Leave only bacteria
metaphlan_PHY_rel_bact <- subset_taxa(metaphlan_PHY_rel, Kingdom != "Viruses" & Kingdom != "Eukaryota" & Kingdom != "Archaea")

# Prune to diff taxa level
metaphlan_PHY_rel_species <- tax_glom(metaphlan_PHY_rel_bact, taxrank = "Species")
metaphlan_PHY_rel_genus <- tax_glom(metaphlan_PHY_rel_bact, taxrank = "Genus")
metaphlan_PHY_rel_family <- tax_glom(metaphlan_PHY_rel_bact, taxrank = "Family")
metaphlan_PHY_rel_order <- tax_glom(metaphlan_PHY_rel_bact, taxrank = "Order")
metaphlan_PHY_rel_class <- tax_glom(metaphlan_PHY_rel_bact, taxrank = "Class")
metaphlan_PHY_rel_phylum <- tax_glom(metaphlan_PHY_rel_bact, taxrank = "Phylum")

```

### MGE Counts

```{r, warning=FALSE, error=FALSE, message=FALSE, results='hide', fig.show='hide'}
counts_MGE <-as.matrix(read.table("MGE_genemat.txt", header= T, check.names = F, row.names = 1))

# Reorder to match metadata
match <- match(rownames(metadata), colnames(counts_MGE))
counts_MGE <- counts_MGE[,match]
all(colnames(counts_MGE) == rownames(metadata))

# Tax table
MGE_tax_table_trim <- read.delim("MGE_tax_table_trim.txt", header=T)
colnames(MGE_tax_table_trim) <- c("Gene", "Element", "Class")

# Reorder tax_table to match
match <- match(rownames(counts_MGE), MGE_tax_table_trim$Gene)
MGE_tax_table_trim <- MGE_tax_table_trim[match,]
all(rownames(counts_MGE) == MGE_tax_table_trim$Gene)

# remove rownames
counts_MGE_phy <- counts_MGE
rownames(counts_MGE_phy) <- c(1:nrow(counts_MGE)+1)

# Combine into phyloseq object
MGE_PHY_counts <- phyloseq(otu_table(counts_MGE_phy, taxa_are_rows=TRUE), 
                       tax_table(as.matrix(MGE_tax_table_trim)), sample_data(metadata))

# Exclude biological / technical replicates
MGE_PHY_counts_uniq <- subset_samples(MGE_PHY_counts, replicate == "NO")
MGE_PHY_counts_uniq <- subset_samples(MGE_PHY_counts_uniq, sample_material == "water")

# Create phyloseq object with only HWW samples
MGE_PHY_counts_HWW <- subset_samples(MGE_PHY_counts_uniq, HWW == "YES")
```

```{r, warning=FALSE, error=FALSE, message=FALSE, results='hide', fig.show='hide'}
counts_MGE <-as.matrix(read.table("cp_MGE_genemat.txt", header= T, check.names = F, row.names = 1))

# Reorder to match metadata
match <- match(rownames(metadata), colnames(counts_MGE))
counts_MGE <- counts_MGE[,match]
all(colnames(counts_MGE) == rownames(metadata))

# Tax table
MGE_tax_table_trim <- read.delim("MGE_tax_table_trim.txt", header=T)
colnames(MGE_tax_table_trim) <- c("Gene", "Element", "Class")

# Reorder tax_table to match
match <- match(rownames(counts_MGE), MGE_tax_table_trim$Gene)
MGE_tax_table_trim <- MGE_tax_table_trim[match,]
all(rownames(counts_MGE) == MGE_tax_table_trim$Gene)

# Normalization to MGE lengths
MGE_lengths <- read.delim("MGE_lengths.txt", 
                          header=FALSE, comment.char="#", check.names = F)
match <- match(rownames(counts_MGE), MGE_lengths$V1)
MGE_lengths <- MGE_lengths[match,]
all(rownames(MGE_tax_table_trim$Gene) == MGE_lengths$V1)
counts_MGE_length_norm <- counts_MGE/MGE_lengths[, 2]

# Normalization with Metaxa2 SSU counts
counts_MGE_length_SSU_norm <- t(t(counts_MGE_length_norm)/metadata$SSU_counts) * 1540
all(rownames(metadata) == colnames(counts_MGE_length_SSU_norm))
identical((counts_MGE_length_norm[1, 5]/metadata$SSU_counts[5]) * 1540, counts_MGE_length_SSU_norm[1, 5])
all(rownames(counts_MGE_length_norm) == MGE_tax_table_trim$Gene)

# Hide rownames
rownames(counts_MGE_length_SSU_norm) <- c(1:nrow(counts_MGE_length_SSU_norm)+1)

dim(MGE_tax_table_trim)
rownames(MGE_tax_table_trim) <- c(1:nrow(counts_MGE_length_SSU_norm)+1)

# Combine to phyloseq object
MGE_rel_PHY <- phyloseq(otu_table(counts_MGE_length_SSU_norm, taxa_are_rows = TRUE), 
                    sample_data(metadata), tax_table(as.matrix(MGE_tax_table_trim)))
```

### Analyses and Plots

#### Resistomes PCA Plot

```{r}
# Plot all samples
a1 <- resfinder_PHY_counts_HWW %>% 
  tax_transform("clr") %>% 
  ord_calc(method = "PCA") %>% 
  ord_plot(color = "hospital", size = 10, alpha = 0.8) +
  theme(text = element_text(family = "Palatino"),
    plot.title = element_text(size = 17, hjust = 0.5),
    axis.title = element_text(size = 17, face = "bold"),
    legend.title = element_blank(),
    legend.text = element_text(size = 17),
    legend.key.size = unit(1, 'cm'),
    legend.position = c(0.15, 0.9),
    legend.background = element_rect(color = "#CECECE"),
    legend.direction = "horizontal",
    axis.text = element_text(size = 17), aspect.ratio = 1) +
  scale_color_manual(values=c("#CD3857", "#61A6E3", "#CC7722")) +
  stat_ellipse(level = 0.95, linetype = 2, aes(colour = hospital), size = 1)

a1_plot <- a1 + plot_annotation(title = "PCA, clr-transformed, for three hospitals on Resistomes", subtitle = "ARGs") & 
  theme(plot.title = element_text(size = 25, family = "Palatino", face = "bold"),
        plot.subtitle = element_text(size = 22, family = "Palatino", hjust = 0.5))

# Save plots for all samples
ggsave(filename = "plot1_arg_pca.tiff",
       width = 16, height = 13, dpi = 300, units = "in", device='tiff', scale = 1)
```

```{r}
a1_load <- resfinder_PHY_counts_HWW %>% 
  tax_transform("clr") %>% 
  ord_calc(method = "PCA") %>% 
  ord_plot(color = "hospital", size = 10, alpha = 0.8, plot_taxa = 1:40, 
           tax_lab_style = tax_lab_style(size=4, alpha=0.8),
           tax_vec_length = 10,
           tax_vec_style_all = vec_tax_all(linewidth = 0.5, alpha = 0.01),
           tax_vec_style_sel = vec_tax_sel(linewidth = 1.2, alpha = 0.8),
           taxon_renamer = function(x) tax_table(resfinder_PHY_counts_HWW)[x,3]) +
  theme(text = element_text(family = "Palatino"),
    plot.title = element_text(size = 17, hjust = 0.5),
    axis.title = element_text(size = 17, face = "bold"),
    legend.title = element_blank(),
    legend.text = element_text(size = 17),
    legend.key.size = unit(1, 'cm'),
    legend.position = c(0.15, 0.9),
    legend.background = element_rect(color = "#CECECE"),
    legend.direction = "horizontal",
    axis.text = element_text(size = 17), aspect.ratio = 1) +
  scale_color_manual(values=c("#CD3857", "#61A6E3", "#CC7722")) +
  stat_ellipse(level = 0.95, linetype = 2, aes(colour = hospital), size = 1)

a1_load_plot <- a1_load + plot_annotation(title = "PCA loading, clr-transformed, for three hospitals on Resistomes", subtitle = "ARGs") & 
  theme(plot.title = element_text(size = 25, family = "Palatino", face = "bold"),
        plot.subtitle = element_text(size = 22, family = "Palatino", hjust = 0.5))

# Save plots for all samples
ggsave(filename = "plot1_arg_pca_load.tiff",
       width = 16, height = 13, dpi = 300, units = "in", device='tiff', scale = 1)
```

```{r}
# Plot all samples
a2 <- resfinder_PHY_counts_HWW %>% 
  tax_transform("clr") %>% 
  ord_calc(method = "PCA") %>% 
  ord_plot(color = "hospital", size = 10, alpha = 0.8) +
  theme(text = element_text(family = "Palatino"),
    plot.title = element_text(size = 17, hjust = 0.5),
    axis.title = element_text(size = 17, face = "bold"),
    legend.title = element_blank(),
    legend.text = element_text(size = 17),
    legend.key.size = unit(1, 'cm'),
    legend.position = c(0.15, 0.9),
    legend.background = element_rect(color = "#CECECE"),
    legend.direction = "horizontal",
    axis.text = element_text(size = 17), aspect.ratio = 1) +
  scale_color_manual(values=c("#CD3857", "#61A6E3", "#CC7722")) +
  stat_ellipse(level = 0.95, linetype = 2, aes(colour = hospital), size = 1)

# Add indication for month with different shapes
a2_plot <- a2 + geom_point(aes(shape = factor(month)), size = 5)  
# Define labels for month legend
month_labels <- c("May", "June", "July")

a2_plot <- a2_plot +
  scale_shape_manual(name = "Month", 
                     values = c(5, 6, 7),  # Assign shapes to each month
                     labels = month_labels) + 
  plot_annotation(title = "PCA, clr-transformed, for three hospitals on Resistomes", subtitle = "ARGs") & 
  theme(plot.title = element_text(size = 25, family = "Palatino", face = "bold"),
        plot.subtitle = element_text(size = 22, family = "Palatino", hjust = 0.5))

# Save plots for all samples
ggsave(filename = "plot2_arg_pca.tiff",
       width = 16, height = 13, dpi = 300, units = "in", device='tiff', scale = 1)
```

#### Permanova Statistics for Resistomes

```{r}
# Test the significance 
resfinder_PHY_counts_uniq_temp <- subset_samples(resfinder_PHY_counts_uniq, (hospital == "SLMC" | hospital == "UST"))
resfinder_PHY_counts_uniq_temp %>%
  dist_calc("aitchison") %>%
  dist_permanova(variables = c("hospital"), n_perms = 9999, seed = 12345)
```

```{r}
# Test the significance 
resfinder_PHY_counts_uniq_temp <- subset_samples(resfinder_PHY_counts_uniq, (hospital == "SLMC" | hospital == "CGH"))
resfinder_PHY_counts_uniq_temp %>%
  dist_calc("aitchison") %>%
  dist_permanova(variables = c("hospital"), n_perms = 9999, seed = 12345)
```

```{r}
# Test the significance 
resfinder_PHY_counts_uniq_temp <- subset_samples(resfinder_PHY_counts_uniq, (hospital == "CGH" | hospital == "UST"))
resfinder_PHY_counts_uniq_temp %>%
  dist_calc("aitchison") %>%
  dist_permanova(variables = c("hospital"), n_perms = 9999, seed = 12345)
```

#### Metaphlan Taxa PCA Plot

```{r}
# Plot all samples
b1 <- metaphlan_PHY_counts_genus %>% 
  tax_transform("clr") %>% 
  ord_calc(method = "PCA") %>% 
  ord_plot(color = "hospital", size = 10, alpha = 0.8) +
  theme(text = element_text(family = "Palatino"),
    plot.title = element_text(size = 17, hjust = 0.5),
    axis.title = element_text(size = 17, face = "bold"),
    legend.title = element_blank(),
    legend.text = element_text(size = 17),
    legend.key.size = unit(1, 'cm'),
    legend.position = c(0.15, 0.9),
    legend.background = element_rect(color = "#CECECE"),
    legend.direction = "horizontal",
    axis.text = element_text(size = 17), aspect.ratio = 1) +
  scale_color_manual(values=c("#CD3857", "#61A6E3", "#CC7722")) +
  stat_ellipse(level = 0.95, linetype = 2, aes(colour = hospital), size = 1)

b1_plot <- b1 + plot_annotation(title = "PCA, clr-transformed, for three hospitals on Taxonomic Composition", subtitle = "Metaphlan") & 
  theme(plot.title = element_text(size = 25, family = "Palatino", face = "bold"),
        plot.subtitle = element_text(size = 22, family = "Palatino", hjust = 0.5))

# Save plots for all samples
ggsave(filename = "plot1_tax_pca.tiff",
       width = 16, height = 13, dpi = 300, units = "in", device='tiff', scale = 1)
```

```{r}
b1_load <- metaphlan_PHY_counts_genus %>% 
  tax_transform("clr") %>% 
  ord_calc(method = "PCA") %>% 
  ord_plot(color = "hospital", size = 10, alpha = 0.8, plot_taxa = 1:40, 
           tax_lab_style = tax_lab_style(size=4, alpha=0.8),
           tax_vec_length = 10,
           tax_vec_style_all = vec_tax_all(linewidth = 0.5, alpha = 0.01),
           tax_vec_style_sel = vec_tax_sel(linewidth = 1.2, alpha = 0.8),
           taxon_renamer = function(x) tax_table(metaphlan_PHY_counts_genus)[x,6]) +
  theme(text = element_text(family = "Palatino"),
    plot.title = element_text(size = 17, hjust = 0.5),
    axis.title = element_text(size = 17, face = "bold"),
    legend.title = element_blank(),
    legend.text = element_text(size = 17),
    legend.key.size = unit(1, 'cm'),
    legend.position = c(0.15, 0.9),
    legend.background = element_rect(color = "#CECECE"),
    legend.direction = "horizontal",
    axis.text = element_text(size = 17), aspect.ratio = 1) +
  scale_color_manual(values=c("#CD3857", "#61A6E3", "#CC7722")) +
  stat_ellipse(level = 0.95, linetype = 2, aes(colour = hospital), size = 1)

b1_load_plot <- b1_load + plot_annotation(title = "PCA loading, clr-transformed, for three hospitals on Taxonomic Composition", subtitle = "Metaphlan") & 
  theme(plot.title = element_text(size = 25, family = "Palatino", face = "bold"),
        plot.subtitle = element_text(size = 22, family = "Palatino", hjust = 0.5))

# Save plots for all samples
ggsave(filename = "plot1_tax_pca_load.tiff",
       width = 16, height = 13, dpi = 300, units = "in", device='tiff', scale = 1)
```

```{r}
# Plot all samples
b2 <- metaphlan_PHY_counts_genus %>% 
  tax_transform("clr") %>% 
  ord_calc(method = "PCA") %>% 
  ord_plot(color = "hospital", size = 10, alpha = 0.8) +
  theme(text = element_text(family = "Palatino"),
    plot.title = element_text(size = 17, hjust = 0.5),
    axis.title = element_text(size = 17, face = "bold"),
    legend.title = element_blank(),
    legend.text = element_text(size = 17),
    legend.key.size = unit(1, 'cm'),
    legend.position = c(0.15, 0.9),
    legend.background = element_rect(color = "#CECECE"),
    legend.direction = "horizontal",
    axis.text = element_text(size = 17), aspect.ratio = 1) +
  scale_color_manual(values=c("#CD3857", "#61A6E3", "#CC7722")) +
  stat_ellipse(level = 0.95, linetype = 2, aes(colour = hospital), size = 1)

# Add indication for month with different shapes
b2_plot <- b2 + geom_point(aes(shape = factor(month)), size = 5)  
# Define labels for month legend
month_labels <- c("May", "June", "July")

b2_plot <- b2_plot +
  scale_shape_manual(name = "Month", 
                     values = c(5, 6, 7),  # Assign shapes to each month
                     labels = month_labels) + 
  plot_annotation(title = "PCA, clr-transformed, for three hospitals on Taxonomic Composition", subtitle = "Metaphlan") & 
  theme(plot.title = element_text(size = 25, family = "Palatino", face = "bold"),
        plot.subtitle = element_text(size = 22, family = "Palatino", hjust = 0.5))

# Save plots for all samples
ggsave(filename = "plot2_tax_pca.tiff",
       width = 16, height = 13, dpi = 300, units = "in", device='tiff', scale = 1)
```

#### Permanova Statistics on Taxa

```{r}
# Test significance
metaphlan_PHY_counts_genus_uniq_temp <- subset_samples(metaphlan_PHY_counts_genus, (hospital == "SLMC" | hospital == "UST"))
metaphlan_PHY_counts_genus_uniq_temp %>%
  dist_calc("aitchison") %>%
  dist_permanova(variables = c("hospital"), n_perms = 9999, seed = 12345)
```

```{r}
# Test significance
metaphlan_PHY_counts_genus_uniq_temp <- subset_samples(metaphlan_PHY_counts_genus, (hospital == "SLMC" | hospital == "CGH"))
metaphlan_PHY_counts_genus_uniq_temp %>%
  dist_calc("aitchison") %>%
  dist_permanova(variables = c("hospital"), n_perms = 9999, seed = 12345)
```

```{r}
# Test significance
metaphlan_PHY_counts_genus_uniq_temp <- subset_samples(metaphlan_PHY_counts_genus, (hospital == "CGH" | hospital == "UST"))
metaphlan_PHY_counts_genus_uniq_temp %>%
  dist_calc("aitchison") %>%
  dist_permanova(variables = c("hospital"), n_perms = 9999, seed = 12345)
```

#### Mobilomes MGE PCA Plot

```{r}
# Plot all samples
c1 <- MGE_PHY_counts_HWW %>% 
  tax_transform("clr") %>% 
  ord_calc(method = "PCA") %>% 
  ord_plot(color = "hospital", size = 10, alpha = 0.8) +
  theme(text = element_text(family = "Palatino"),
    plot.title = element_text(size = 17, hjust = 0.5),
    axis.title = element_text(size = 17, face = "bold"),
    legend.title = element_blank(),
    legend.text = element_text(size = 17),
    legend.key.size = unit(1, 'cm'),
    legend.position = c(0.8, 0.9),
    legend.background = element_rect(color = "#CECECE"),
    legend.direction = "horizontal",
    axis.text = element_text(size = 17), aspect.ratio = 1) +
  scale_color_manual(values=c("#CD3857", "#61A6E3", "#CC7722")) +
  stat_ellipse(level = 0.95, linetype = 2, aes(colour = hospital), size = 1)

c1_plot <- c1 + plot_annotation(title = "PCA, clr-transformed, for three hospitals on Mobilomes", subtitle = "MGEs") & 
  theme(plot.title = element_text(size = 25, family = "Palatino", face = "bold"),
        plot.subtitle = element_text(size = 22, family = "Palatino", hjust = 0.5))

# Save plots for all samples
ggsave(filename = "plot1_mge_pca.tiff",
       width = 16, height = 13, dpi = 300, units = "in", device='tiff', scale = 1)
```

```{r}
c1_load <- MGE_PHY_counts_HWW %>% 
  tax_transform("clr") %>% 
  ord_calc(method = "PCA") %>% 
  ord_plot(color = "hospital", size = 10, alpha = 0.8, plot_taxa = 1:40, 
           tax_lab_style = tax_lab_style(size=4, alpha=0.8),
           tax_vec_length = 10,
           tax_vec_style_all = vec_tax_all(linewidth = 0.5, alpha = 0.01),
           tax_vec_style_sel = vec_tax_sel(linewidth = 1.2, alpha = 0.8),
           taxon_renamer = function(x) tax_table(MGE_PHY_counts_HWW)[x,1]) +
  theme(text = element_text(family = "Palatino"),
    plot.title = element_text(size = 17, hjust = 0.5),
    axis.title = element_text(size = 17, face = "bold"),
    legend.title = element_blank(),
    legend.text = element_text(size = 17),
    legend.key.size = unit(1, 'cm'),
    legend.position = c(0.15, 0.9),
    legend.background = element_rect(color = "#CECECE"),
    legend.direction = "horizontal",
    axis.text = element_text(size = 17), aspect.ratio = 1) +
  scale_color_manual(values=c("#CD3857", "#61A6E3", "#CC7722")) +
  stat_ellipse(level = 0.95, linetype = 2, aes(colour = hospital), size = 1)

c1_load_plot <- c1_load + plot_annotation(title = "PCA loading, clr-transformed, for three hospitals on Mobilomes", subtitle = "MGEs") & 
  theme(plot.title = element_text(size = 25, family = "Palatino", face = "bold"),
        plot.subtitle = element_text(size = 22, family = "Palatino", hjust = 0.5))

# Save plots for all samples
ggsave(filename = "plot1_mge_pca_load.tiff",
       width = 16, height = 13, dpi = 300, units = "in", device='tiff', scale = 1)
```

```{r}
# Plot all samples
c2 <- MGE_PHY_counts_HWW %>% 
  tax_transform("clr") %>% 
  ord_calc(method = "PCA") %>% 
  ord_plot(color = "hospital", size = 10, alpha = 0.8) +
  theme(text = element_text(family = "Palatino"),
    plot.title = element_text(size = 17, hjust = 0.5),
    axis.title = element_text(size = 17, face = "bold"),
    legend.title = element_blank(),
    legend.text = element_text(size = 17),
    legend.key.size = unit(1, 'cm'),
    legend.position = c(0.8, 0.9),
    legend.background = element_rect(color = "#CECECE"),
    legend.direction = "horizontal",
    axis.text = element_text(size = 17), aspect.ratio = 1) +
  scale_color_manual(values=c("#CD3857", "#61A6E3", "#CC7722")) +
  stat_ellipse(level = 0.95, linetype = 2, aes(colour = hospital), size = 1)

# Add indication for month with different shapes
c2_plot <- c2 + geom_point(aes(shape = factor(month)), size = 5)  
# Define labels for month legend
month_labels <- c("May", "June", "July")

c2_plot <- c2_plot +
  scale_shape_manual(name = "Month", 
                     values = c(5, 6, 7),  # Assign shapes to each month
                     labels = month_labels) +
  plot_annotation(title = "PCA, clr-transformed, for three hospitals on Mobilomes", subtitle = "MGEs") & 
  theme(plot.title = element_text(size = 25, family = "Palatino", face = "bold"),
        plot.subtitle = element_text(size = 22, family = "Palatino", hjust = 0.5))

# Save plots for all samples
ggsave(filename = "plot2_mge_pca.tiff",
       width = 16, height = 13, dpi = 300, units = "in", device='tiff', scale = 1)
```

#### Permanova Statistics for MGE

```{r}
MGE_PHY_counts_uniq_temp <- subset_samples(MGE_PHY_counts_uniq, (hospital == "SLMC" | hospital == "UST"))
MGE_PHY_counts_uniq_temp %>%
  dist_calc("aitchison") %>%
  dist_permanova(variables = c("hospital"), n_perms = 9999, seed = 12345)
```

```{r}
MGE_PHY_counts_uniq_temp <- subset_samples(MGE_PHY_counts_uniq, (hospital == "SLMC" | hospital == "CGH"))
MGE_PHY_counts_uniq_temp %>%
  dist_calc("aitchison") %>%
  dist_permanova(variables = c("hospital"), n_perms = 9999, seed = 12345)
```

```{r}
MGE_PHY_counts_uniq_temp <- subset_samples(MGE_PHY_counts_uniq, (hospital == "CGH" | hospital == "UST"))
MGE_PHY_counts_uniq_temp %>%
  dist_calc("aitchison") %>%
  dist_permanova(variables = c("hospital"), n_perms = 9999, seed = 12345)

```

#### PCA Plot for Resistomes, Taxa and Mobilomes

```{r}
patchwork0 <- a1_plot + b1_plot + c1_plot + 
  plot_annotation(tag_levels = list(c('ARGs', 'Taxa', 'MGEs')), 
  title = "PCA, clr-transformed, for Resistomes, Taxonomic Composition and Mobilomes") & 
  theme(legend.position = "bottom",
        plot.title = element_text(size = 30, family = "Palatino", face = "bold"),
        plot.tag = element_text(size = 22, face = "bold"),
        plot.tag.position = c(0.5, 1))
patchwork <- patchwork0 + plot_layout(guides = "collect")
patchwork

# Save plots for HWW samples
ggsave(filename = "plot1a_merged_pca.tiff",
       width = 16, height = 13, dpi = 300, units = "in", device='tiff', scale = 1)
ggsave(filename = "plot1b_merged_pca.tiff",
       width = 32, height = 13, dpi = 300, units = "in", device='tiff', scale = 1)
```

```{r}
patchwork0 <- a2_plot + b2_plot + c2_plot + 
  plot_annotation(tag_levels = list(c('ARGs', 'Taxa', 'MGEs')), 
  title = "PCA, clr-transformed, for Resistomes, Taxonomic Composition and Mobilomes") & 
  theme(legend.position = "bottom",
        plot.title = element_text(size = 30, family = "Palatino", face = "bold"),
        plot.tag = element_text(size = 22, face = "bold"),
        plot.tag.position = c(0.5, 1))
patchwork <- patchwork0 + plot_layout(guides = "collect")
patchwork

# Save plots for HWW samples
ggsave(filename = "plot2a_merged_pca.tiff",
       width = 16, height = 13, dpi = 300, units = "in", device='tiff', scale = 1)
ggsave(filename = "plot2b_merged_pca.tiff",
       width = 32, height = 13, dpi = 300, units = "in", device='tiff', scale = 1)

```

### Diversity Metrics

#### Shannon for ARGs

```{r, warning=FALSE, error=FALSE, message=FALSE, results='hide', fig.show='hide'}
arg_shannon_div <- data.frame("shannon" = diversity(otu_table(resfinder_PHY_counts_HWW), index = "shannon", MARGIN = 2))
all(rownames(arg_shannon_div) == colnames(resfinder_PHY_counts_HWW@otu_table))
diversity_observed <- apply(resfinder_PHY_counts_HWW@otu_table > 0, 2, sum)
arg_shannon_div$observed <- diversity_observed
arg_shannon_div$hospital <- as.factor(resfinder_PHY_counts_HWW@sam_data$hospital)

# Get mean and standard deviation
arg_shannon_eda <- do.call(data.frame, aggregate(shannon ~ hospital, arg_shannon_div, function(x) c(mean = mean(x), sd = sd(x))))

# Test the significance
kruskal.test(shannon ~ hospital, data = arg_shannon_div)

# There are no sig. differences

# Test multiple pairwise
arg_shannon_stat.test <- pairwise.wilcox.test(arg_shannon_div$shannon, arg_shannon_div$hospital,
                 p.adjust.method = "BH")
arg_shannon_stat.test

# Plot
p <- ggplot(arg_shannon_div, aes(x=hospital, y=shannon, fill = hospital)) + 
  geom_boxplot()

# Add ID as label to the plot
p1 <- p + geom_dotplot(binaxis='y', stackdir='center', dotsize = 0.4) + 
  scale_fill_manual(values=c("#CD3857", "#61A6E3", "#CC7722")) +
  theme_classic() + theme(text = element_text(family = "Palatino"),
                          axis.title.y = element_text(size = 35),
                          axis.title.x = element_blank(),
                          axis.text.y = element_text(size = 20),
                          axis.text.x = element_text(size = 16, angle = 30, hjust = 1, face = "bold"),
                          legend.position = "none",
                          plot.title = element_text(face = "bold"),
                          plot.margin = margin(0.25, 0.25, 0.25, 0.25, "cm"))

# Adjust the scale to be more detailed
p1 <- p1 + labs(title = "Diversity Metrics for ARGs Across Hospitals", 
                subtitle = "Shannon Diversity Index by Hospital") +
                scale_y_continuous(limits = c(4,max(arg_shannon_div$shannon)), breaks = seq(4,max(arg_shannon_div$shannon), by = 0.5))

# Add ID as labels
p1 <- p1 + geom_text(aes(label = paste(rownames(arg_shannon_div), arg_shannon_div$shannon)), hjust = -0.1, vjust = 0.5, size = 4)

p1  # Display the plot

# Save plot
ggsave(filename = "plot5_arg_shannon.tiff",
       width = 16, height = 13, dpi = 300, units = "in", device='tiff', scale = 1)
```

#### Simpson for ARGs

```{r, warning=FALSE, error=FALSE, message=FALSE, results='hide', fig.show='hide'}
arg_simpson_div <- data.frame("simpson" = diversity(otu_table(resfinder_PHY_counts_HWW), index = "simpson", MARGIN = 2))
all(rownames(arg_simpson_div) == colnames(resfinder_PHY_counts_HWW@otu_table))
diversity_observed <- apply(resfinder_PHY_counts_HWW@otu_table > 0, 2, sum)
arg_simpson_div$observed <- diversity_observed
arg_simpson_div$hospital <- as.factor(resfinder_PHY_counts_HWW@sam_data$hospital)

# Get mean and standard deviation
arg_simpson_eda <- do.call(data.frame, aggregate(simpson ~ hospital, arg_simpson_div, function(x) c(mean = mean(x), sd = sd(x))))

# Test the significance
kruskal.test(simpson ~ hospital, data = arg_simpson_div)

# There are no sig. differences

# Test multiple pairwise
arg_simpson_stat.test <- pairwise.wilcox.test(arg_simpson_div$simpson, arg_simpson_div$hospital,
                 p.adjust.method = "BH")
arg_simpson_stat.test

# Plot
p <- ggplot(arg_simpson_div, aes(x=hospital, y=simpson, fill = hospital)) + 
  geom_boxplot()

# Add ID as label to the plot
p1 <- p + geom_dotplot(binaxis='y', stackdir='center', dotsize = 0.4) + 
  scale_fill_manual(values=c("#CD3857", "#61A6E3", "#CC7722")) +
  theme_classic() + theme(text = element_text(family = "Palatino"),
                          axis.title.y = element_text(size = 35),
                          axis.title.x = element_blank(),
                          axis.text.y = element_text(size = 20),
                          axis.text.x = element_text(size = 16, angle = 30, hjust = 1, face = "bold"),
                          legend.position = "none",
                          plot.title = element_text(face = "bold"),
                          plot.margin = margin(0.25, 0.25, 0.25, 0.25, "cm"))

# Adjust the scale to be more detailed
p1 <- p1 + labs(title = "Diversity Metrics for ARGs Across Hospitals", 
                subtitle = "Simpson Diversity Index by Hospital") +
                scale_y_continuous(limits = c(0.94,max(arg_simpson_div$simpson)), breaks = seq(0.94,max(arg_simpson_div$simpson), by = 0.02))

# Add ID as labels
p1 <- p1 + geom_text(aes(label = paste(rownames(arg_simpson_div), arg_simpson_div$simpson)), hjust = -0.1, vjust = 0.5, size = 4)

p1  # Display the plot

# Save plot
ggsave(filename = "plot5_arg_simpson.tiff",
       width = 16, height = 13, dpi = 300, units = "in", device='tiff', scale = 1)
```
#### Inverse Simpson for ARGs

```{r, warning=FALSE, error=FALSE, message=FALSE, results='hide', fig.show='hide'}
arg_invsimp_div <- data.frame("invsimpson" = diversity(otu_table(resfinder_PHY_counts_HWW), index = "inv", MARGIN = 2))
all(rownames(arg_invsimp_div) == colnames(resfinder_PHY_counts_HWW@otu_table))
diversity_observed <- apply(resfinder_PHY_counts_HWW@otu_table > 0, 2, sum)
arg_invsimp_div$observed <- diversity_observed
arg_invsimp_div$hospital <- as.factor(resfinder_PHY_counts_HWW@sam_data$hospital)

# Get mean and standard deviation
arg_invsimp_eda <- do.call(data.frame, aggregate(invsimpson ~ hospital, arg_invsimp_div, function(x) c(mean = mean(x), sd = sd(x))))

# Test the significance
kruskal.test(invsimpson ~ hospital, data = arg_invsimp_div)

# There are no sig. differences

# Test multiple pairwise
arg_invsimp_stat.test <- pairwise.wilcox.test(arg_invsimp_div$invsimpson, arg_invsimp_div$hospital,
                 p.adjust.method = "BH")
arg_invsimp_stat.test

# Plot
p <- ggplot(arg_invsimp_div, aes(x=hospital, y=invsimpson, fill = hospital)) + 
  geom_boxplot()

# Add ID as label to the plot
p1 <- p + geom_dotplot(binaxis='y', stackdir='center', dotsize = 0.4) + 
  scale_fill_manual(values=c("#CD3857", "#61A6E3", "#CC7722")) +
  theme_classic() + theme(text = element_text(family = "Palatino"),
                          axis.title.y = element_text(size = 35),
                          axis.title.x = element_blank(),
                          axis.text.y = element_text(size = 20),
                          axis.text.x = element_text(size = 16, angle = 30, hjust = 1, face = "bold"),
                          legend.position = "none",
                          plot.title = element_text(face = "bold"),
                          plot.margin = margin(0.25, 0.25, 0.25, 0.25, "cm"))

# Adjust the scale to be more detailed
p1 <- p1 + labs(title = "Diversity Metrics for ARGs Across Hospitals", 
                subtitle = "Inv Simpson Diversity Index by Hospital") +
                scale_y_continuous(limits = c(15,max(arg_invsimp_div$invsimpson)), breaks = seq(15,max(arg_invsimp_div$invsimpson), by = 10))

# Add ID as labels
p1 <- p1 + geom_text(aes(label = paste(rownames(arg_invsimp_div), arg_invsimp_div$invsimpson)), hjust = -0.1, vjust = 0.5, size = 4)

p1  # Display the plot

# Save plot
ggsave(filename = "plot5_arg_invsimpson.tiff",
       width = 16, height = 13, dpi = 300, units = "in", device='tiff', scale = 1)
```
#### Richness for ARGs

```{r, warning=FALSE, error=FALSE, message=FALSE, results='hide', fig.show='hide'}
arg_rich <- data.frame("richness" = specnumber(otu_table(resfinder_PHY_counts_HWW), MARGIN=2))
all(rownames(arg_rich) == colnames(resfinder_PHY_counts_HWW@otu_table))
arg_rich$hospital <- as.factor(resfinder_PHY_counts_HWW@sam_data$hospital)

# Get mean and standard deviation
arg_rich_eda <- do.call(data.frame, aggregate(richness ~ hospital, arg_rich, function(x) c(mean = mean(x), sd = sd(x))))

# Test the significance
kruskal.test(richness ~ hospital, data = arg_rich)

# There are sig. differences

# Test multiple pairwise
arg_rich_stat.test <- pairwise.wilcox.test(arg_rich$richness, arg_rich$hospital,
                 p.adjust.method = "BH")
arg_rich_stat.test

pvalues <- data.frame(group1 = c("SLMC","UST","CGH"), group2 = c("CGH","SLMC","UST"), p.adj = c("**","n.s.","**"))

# Plot
p <- ggboxplot(arg_rich, x = "hospital", y = "richness", fill = "hospital", palette = c("#CD3857", "#61A6E3", "#CC7722")) + 
  stat_pvalue_manual(pvalues, size = 6, y.position = 800, step.increase = 0.05, label = "p.adj")

# Add ID as label to the plot
p1 <- p + geom_dotplot(binaxis='y', stackdir='center', dotsize = 0.4, aes(fill=hospital)) + 
  scale_fill_manual(values=c("#CD3857", "#61A6E3", "#CC7722")) +
  theme_classic() + theme(text = element_text(family = "Palatino"),
                          axis.title.y = element_text(size = 35),
                          axis.title.x = element_blank(),
                          axis.text.y = element_text(size = 20),
                          axis.text.x = element_text(size = 16, angle = 30, hjust = 1, face = "bold"),
                          legend.position = "none",
                          plot.title = element_text(face = "bold"),
                          plot.margin = margin(0.25, 0.25, 0.25, 0.25, "cm"))

# Adjust the scale to be more detailed
p1 <- p1 + labs(title = "Richness for ARGs Across Hospitals", 
                subtitle = "Richness by Hospital") +
                scale_y_continuous(limits = c(350,850), breaks = seq(350,850, by = 50))

# Add ID as labels
p1 <- p1 + geom_text(aes(label = paste(rownames(arg_rich), arg_rich$richness)), hjust = -0.1, vjust = 0.5, size = 4)

p1  # Display the plot

# Save plot
ggsave(filename = "plot5_arg_richness.tiff",
       width = 16, height = 13, dpi = 300, units = "in", device='tiff', scale = 1)
```
#### Evenness for ARGs

```{r, warning=FALSE, error=FALSE, message=FALSE, results='hide', fig.show='hide'}
arg_even <- data.frame("evenness" = arg_shannon_div$shannon / log(arg_rich$richness))
rownames(arg_even) <- colnames(resfinder_PHY_counts_HWW@otu_table)
all(rownames(arg_even) == colnames(resfinder_PHY_counts_HWW@otu_table))
arg_even$hospital <- as.factor(resfinder_PHY_counts_HWW@sam_data$hospital)

# Get mean and standard deviation
arg_even_eda <- do.call(data.frame, aggregate(evenness ~ hospital, arg_even, function(x) c(mean = mean(x), sd = sd(x))))


# Test the significance
kruskal.test(evenness ~ hospital, data = arg_even)

# There are sig. differences

# Test multiple pairwise
arg_even_stat.test <- pairwise.wilcox.test(arg_even$evenness, arg_even$hospital,
                 p.adjust.method = "BH")
arg_even_stat.test

pvalues <- data.frame(group1 = c("SLMC","UST","CGH"), group2 = c("CGH","SLMC","UST"), p.adj = c("n.s.","n.s.","**"))

# Plot
p <- ggboxplot(arg_even, x = "hospital", y = "evenness", fill = "hospital", palette = c("#CD3857", "#61A6E3", "#CC7722")) + 
  stat_pvalue_manual(pvalues, size = 6, y.position = 0.85, step.increase = 0.05, label = "p.adj")

# Add ID as label to the plot
p1 <- p + geom_dotplot(binaxis='y', stackdir='center', dotsize = 0.4, aes(fill=hospital)) + 
  scale_fill_manual(values=c("#CD3857", "#61A6E3", "#CC7722")) +
  theme_classic() + theme(text = element_text(family = "Palatino"),
                          axis.title.y = element_text(size = 35),
                          axis.title.x = element_blank(),
                          axis.text.y = element_text(size = 20),
                          axis.text.x = element_text(size = 16, angle = 30, hjust = 1, face = "bold"),
                          legend.position = "none",
                          plot.title = element_text(face = "bold"),
                          plot.margin = margin(0.25, 0.25, 0.25, 0.25, "cm"))

# Adjust the scale to be more detailed
p1 <- p1 + labs(title = "Evenness for ARGs Across Hospitals", 
                subtitle = "Pielou's Evenness by Hospital") +
                scale_y_continuous(limits = c(0.65,0.9), breaks = seq(0.65,0.9, by = 0.05))

# Add ID as labels
p1 <- p1 + geom_text(aes(label = paste(rownames(arg_even), arg_even$evenness)), hjust = -0.1, vjust = 0.5, size = 4)

p1  # Display the plot

# Save plot
ggsave(filename = "plot5_arg_evenness.tiff",
       width = 16, height = 13, dpi = 300, units = "in", device='tiff', scale = 1)
```

#### Shannon for Taxa (metaphlan)

```{r, warning=FALSE, error=FALSE, message=FALSE, results='hide', fig.show='hide'}
tax_shannon_div <- data.frame("shannon" = diversity(otu_table(metaphlan_PHY_counts_genus), index = "shannon", MARGIN = 2))
all(rownames(tax_shannon_div) == colnames(metaphlan_PHY_counts_genus@otu_table))
diversity_observed <- apply(metaphlan_PHY_counts_genus@otu_table > 0, 2, sum)
tax_shannon_div$observed <- diversity_observed
tax_shannon_div$hospital <- as.factor(metaphlan_PHY_counts_genus@sam_data$hospital)

# Get mean and standard deviation
tax_shannon_eda <- do.call(data.frame, aggregate(shannon ~ hospital, tax_shannon_div, function(x) c(mean = mean(x), sd = sd(x))))

# Test the significance
kruskal.test(shannon ~ hospital, data = tax_shannon_div)

# There are no sig. differences

# Test multiple pairwise
tax_shannon_stat.test <- pairwise.wilcox.test(tax_shannon_div$shannon, tax_shannon_div$hospital,
                 p.adjust.method = "BH")
tax_shannon_stat.test

# Plot
p <- ggplot(tax_shannon_div, aes(x=hospital, y=shannon, fill = hospital)) + 
  geom_boxplot()

# Add ID as label to the plot
p1 <- p + geom_dotplot(binaxis='y', stackdir='center', dotsize = 0.4) + 
  scale_fill_manual(values=c("#CD3857", "#61A6E3", "#CC7722")) +
  theme_classic() + theme(text = element_text(family = "Palatino"),
                          axis.title.y = element_text(size = 35),
                          axis.title.x = element_blank(),
                          axis.text.y = element_text(size = 20),
                          axis.text.x = element_text(size = 16, angle = 30, hjust = 1, face = "bold"),
                          legend.position = "none",
                          plot.title = element_text(face = "bold"),
                          plot.margin = margin(0.25, 0.25, 0.25, 0.25, "cm"))

# Adjust the scale to be more detailed
p1 <- p1 + labs(title = "Diversity Metrics for Taxonomic Composition Across Hospitals", 
                subtitle = "Shannon Diversity Index by Hospital") +
                scale_y_continuous(limits = c(2.5,max(tax_shannon_div$shannon)), breaks = seq(2.5,max(tax_shannon_div$shannon), by = 0.5))

# Add ID as labels
p1 <- p1 + geom_text(aes(label = paste(rownames(tax_shannon_div), tax_shannon_div$shannon)), hjust = -0.1, vjust = 0.5, size = 4)

p1  # Display the plot

# Save plot
ggsave(filename = "plot6_tax_shannon.tiff",
       width = 16, height = 13, dpi = 300, units = "in", device='tiff', scale = 1)
```

#### Simpson for Taxa (metaphlan)

```{r, warning=FALSE, error=FALSE, message=FALSE, results='hide', fig.show='hide'}
tax_simpson_div <- data.frame("simpson" = diversity(otu_table(metaphlan_PHY_counts_genus), index = "simpson", MARGIN = 2))
all(rownames(tax_simpson_div) == colnames(metaphlan_PHY_counts_genus@otu_table))
diversity_observed <- apply(metaphlan_PHY_counts_genus@otu_table > 0, 2, sum)
tax_simpson_div$observed <- diversity_observed
tax_simpson_div$hospital <- as.factor(metaphlan_PHY_counts_genus@sam_data$hospital)

# Get mean and standard deviation
tax_simpson_eda <- do.call(data.frame, aggregate(simpson ~ hospital, tax_simpson_div, function(x) c(mean = mean(x), sd = sd(x))))

# Test the significance
kruskal.test(simpson ~ hospital, data = tax_simpson_div)

# There are no sig. differences

# Test multiple pairwise
tax_simpson_stat.test <- pairwise.wilcox.test(tax_simpson_div$simpson, tax_simpson_div$hospital,
                 p.adjust.method = "BH")
tax_simpson_stat.test

# Plot
p <- ggplot(tax_simpson_div, aes(x=hospital, y=simpson, fill = hospital)) + 
  geom_boxplot()

# Add ID as label to the plot
p1 <- p + geom_dotplot(binaxis='y', stackdir='center', dotsize = 0.4) + 
  scale_fill_manual(values=c("#CD3857", "#61A6E3", "#CC7722")) +
  theme_classic() + theme(text = element_text(family = "Palatino"),
                          axis.title.y = element_text(size = 35),
                          axis.title.x = element_blank(),
                          axis.text.y = element_text(size = 20),
                          axis.text.x = element_text(size = 16, angle = 30, hjust = 1, face = "bold"),
                          legend.position = "none",
                          plot.title = element_text(face = "bold"),
                          plot.margin = margin(0.25, 0.25, 0.25, 0.25, "cm"))

# Adjust the scale to be more detailed
p1 <- p1 + labs(title = "Diversity Metrics for Taxonomic Composition Across Hospitals", 
                subtitle = "Simpson Diversity Index by Hospital") +
                scale_y_continuous(limits = c(0.9,max(tax_simpson_div$simpson)), breaks = seq(0.9,max(tax_simpson_div$simpson), by = 0.02))

# Add ID as labels
p1 <- p1 + geom_text(aes(label = paste(rownames(tax_simpson_div), tax_simpson_div$simpson)), hjust = -0.1, vjust = 0.5, size = 4)

p1  # Display the plot

# Save plot
ggsave(filename = "plot6_tax_simpson.tiff",
       width = 16, height = 13, dpi = 300, units = "in", device='tiff', scale = 1)
```

#### Inverse Simpson for Taxa (metaphlan)

```{r, warning=FALSE, error=FALSE, message=FALSE, results='hide', fig.show='hide'}
tax_invsimp_div <- data.frame("invsimpson" = diversity(otu_table(metaphlan_PHY_counts_genus), index = "inv", MARGIN = 2))
all(rownames(tax_invsimp_div) == colnames(metaphlan_PHY_counts_genus@otu_table))
diversity_observed <- apply(metaphlan_PHY_counts_genus@otu_table > 0, 2, sum)
tax_invsimp_div$observed <- diversity_observed
tax_invsimp_div$hospital <- as.factor(metaphlan_PHY_counts_genus@sam_data$hospital)

# Get mean and standard deviation
tax_invsimp_eda <- do.call(data.frame, aggregate(invsimpson ~ hospital, tax_invsimp_div, function(x) c(mean = mean(x), sd = sd(x))))

# Test the significance
kruskal.test(invsimpson ~ hospital, data = tax_invsimp_div)

# There are no sig. differences

# Test multiple pairwise
tax_invsimp_stat.test <- pairwise.wilcox.test(tax_invsimp_div$invsimpson, tax_invsimp_div$hospital,
                 p.adjust.method = "BH")
tax_invsimp_stat.test

# Plot
p <- ggplot(tax_invsimp_div, aes(x=hospital, y=invsimpson, fill = hospital)) + 
  geom_boxplot()

# Add ID as label to the plot
p1 <- p + geom_dotplot(binaxis='y', stackdir='center', dotsize = 0.4) + 
  scale_fill_manual(values=c("#CD3857", "#61A6E3", "#CC7722")) +
  theme_classic() + theme(text = element_text(family = "Palatino"),
                          axis.title.y = element_text(size = 35),
                          axis.title.x = element_blank(),
                          axis.text.y = element_text(size = 20),
                          axis.text.x = element_text(size = 16, angle = 30, hjust = 1, face = "bold"),
                          legend.position = "none",
                          plot.title = element_text(face = "bold"),
                          plot.margin = margin(0.25, 0.25, 0.25, 0.25, "cm"))

# Adjust the scale to be more detailed
p1 <- p1 + labs(title = "Diversity Metrics for Taxonomic Composition Across Hospitals", 
                subtitle = "Inv Simpson Diversity Index by Hospital") +
                scale_y_continuous(limits = c(10,max(tax_invsimp_div$invsimpson)), breaks = seq(10,max(tax_invsimp_div$invsimpson), by = 5))

# Add ID as labels
p1 <- p1 + geom_text(aes(label = paste(rownames(tax_invsimp_div), tax_invsimp_div$invsimpson)), hjust = -0.1, vjust = 0.5, size = 4)

p1  # Display the plot

# Save plot
ggsave(filename = "plot6_tax_invsimpson.tiff",
       width = 16, height = 13, dpi = 300, units = "in", device='tiff', scale = 1)
```

#### Richness for Taxa (metaphlan)

```{r, warning=FALSE, error=FALSE, message=FALSE, results='hide', fig.show='hide'}
tax_rich <- data.frame("richness" = specnumber(otu_table(metaphlan_PHY_counts_genus), MARGIN=2))
all(rownames(tax_rich) == colnames(metaphlan_PHY_counts_genus@otu_table))
tax_rich$hospital <- as.factor(metaphlan_PHY_counts_genus@sam_data$hospital)

# Get mean and standard deviation
tax_rich_eda <- do.call(data.frame, aggregate(richness ~ hospital, tax_rich, function(x) c(mean = mean(x), sd = sd(x))))

# Test the significance
kruskal.test(richness ~ hospital, data = tax_rich)

# There are no sig. differences

# Test multiple pairwise
tax_rich_stat.test <- pairwise.wilcox.test(tax_rich$richness, tax_rich$hospital,
                 p.adjust.method = "BH")
tax_rich_stat.test

# Plot
p <- ggplot(tax_rich, aes(x=hospital, y=richness, fill = hospital)) + 
  geom_boxplot()

# Add ID as label to the plot
p1 <- p + geom_dotplot(binaxis='y', stackdir='center', dotsize = 0.4) + 
  scale_fill_manual(values=c("#CD3857", "#61A6E3", "#CC7722")) +
  theme_classic() + theme(text = element_text(family = "Palatino"),
                          axis.title.y = element_text(size = 35),
                          axis.title.x = element_blank(),
                          axis.text.y = element_text(size = 20),
                          axis.text.x = element_text(size = 16, angle = 30, hjust = 1, face = "bold"),
                          legend.position = "none",
                          plot.title = element_text(face = "bold"),
                          plot.margin = margin(0.25, 0.25, 0.25, 0.25, "cm"))

# Adjust the scale to be more detailed
p1 <- p1 + labs(title = "Richness for Taxonomic Composition Across Hospitals", 
                subtitle = "Richness by Hospital") +
                scale_y_continuous(limits = c(200,max(tax_rich$richness)), breaks = seq(200,max(tax_rich$richness), by = 50))

# Add ID as labels
p1 <- p1 + geom_text(aes(label = paste(rownames(tax_rich), tax_rich$richness)), hjust = -0.1, vjust = 0.5, size = 4)

p1  # Display the plot

# Save plot
ggsave(filename = "plot6_tax_richness.tiff",
       width = 16, height = 13, dpi = 300, units = "in", device='tiff', scale = 1)
```

#### Evenness for Taxa (metaphlan)

```{r, warning=FALSE, error=FALSE, message=FALSE, results='hide', fig.show='hide'}
tax_even <- data.frame("evenness" = tax_shannon_div$shannon / log(tax_rich$richness))
rownames(tax_even) <- colnames(metaphlan_PHY_counts_genus@otu_table)
all(rownames(tax_even) == colnames(metaphlan_PHY_counts_genus@otu_table))
tax_even$hospital <- as.factor(metaphlan_PHY_counts_genus@sam_data$hospital)

# Get mean and standard deviation
tax_even_eda <- do.call(data.frame, aggregate(evenness ~ hospital, tax_even, function(x) c(mean = mean(x), sd = sd(x))))

# Test the significance
kruskal.test(evenness ~ hospital, data = tax_even)

# There are no sig. differences

# Test multiple pairwise
tax_even_stat.test <- pairwise.wilcox.test(tax_even$evenness, tax_even$hospital,
                 p.adjust.method = "BH")
tax_even_stat.test

# Plot
p <- ggplot(tax_even, aes(x=hospital, y=evenness, fill = hospital)) + 
  geom_boxplot()

# Add ID as label to the plot
p1 <- p + geom_dotplot(binaxis='y', stackdir='center', dotsize = 0.4) + 
  scale_fill_manual(values=c("#CD3857", "#61A6E3", "#CC7722")) +
  theme_classic() + theme(text = element_text(family = "Palatino"),
                          axis.title.y = element_text(size = 35),
                          axis.title.x = element_blank(),
                          axis.text.y = element_text(size = 20),
                          axis.text.x = element_text(size = 16, angle = 30, hjust = 1, face = "bold"),
                          legend.position = "none",
                          plot.title = element_text(face = "bold"),
                          plot.margin = margin(0.25, 0.25, 0.25, 0.25, "cm"))

# Adjust the scale to be more detailed
p1 <- p1 + labs(title = "Evenness for Taxonomic Composition Across Hospitals", 
                subtitle = "Pielou's Evenness by Hospital") +
                scale_y_continuous(limits = c(0.5,max(tax_even$evenness)), breaks = seq(0.5,max(tax_even$evenness), by = 0.05))

# Add ID as labels
p1 <- p1 + geom_text(aes(label = paste(rownames(tax_even), tax_even$evenness)), hjust = -0.1, vjust = 0.5, size = 4)

p1  # Display the plot

# Save plot
ggsave(filename = "plot6_tax_evenness.tiff",
       width = 16, height = 13, dpi = 300, units = "in", device='tiff', scale = 1)
```

#### Shannon for MGEs

```{r, warning=FALSE, error=FALSE, message=FALSE, results='hide', fig.show='hide'}
mge_shannon_div <- data.frame("shannon" = diversity(otu_table(MGE_PHY_counts_HWW), index = "shannon", MARGIN = 2))
all(rownames(mge_shannon_div) == colnames(MGE_PHY_counts_HWW@otu_table))
diversity_observed <- apply(MGE_PHY_counts_HWW@otu_table > 0, 2, sum)
mge_shannon_div$observed <- diversity_observed
mge_shannon_div$hospital <- as.factor(MGE_PHY_counts_HWW@sam_data$hospital)

# Get mean and standard deviation
mge_shannon_eda <- do.call(data.frame, aggregate(shannon ~ hospital, mge_shannon_div, function(x) c(mean = mean(x), sd = sd(x))))

# Plot
p <- ggplot(mge_shannon_div, aes(x=hospital, y=shannon, fill = hospital)) + 
  geom_boxplot()

# Test the significance
kruskal.test(shannon ~ hospital, data = mge_shannon_div)

# There are no sig. differences

# Test multiple pairwise
mge_shannon_stat.test <- pairwise.wilcox.test(mge_shannon_div$shannon, mge_shannon_div$hospital,
                 p.adjust.method = "BH")
mge_shannon_stat.test

# Add ID as label to the plot
p1 <- p + geom_dotplot(binaxis='y', stackdir='center', dotsize = 0.4) + 
  scale_fill_manual(values=c("#CD3857", "#61A6E3", "#CC7722")) +
  theme_classic() + theme(text = element_text(family = "Palatino"),
                          axis.title.y = element_text(size = 35),
                          axis.title.x = element_blank(),
                          axis.text.y = element_text(size = 20),
                          axis.text.x = element_text(size = 16, angle = 30, hjust = 1, face = "bold"),
                          legend.position = "none",
                          plot.title = element_text(face = "bold"),
                          plot.margin = margin(0.25, 0.25, 0.25, 0.25, "cm"))

# Adjust the scale to be more detailed
p1 <- p1 + labs(title = "Diversity Metrics for MGEs Across Hospitals", 
                subtitle = "Shannon Diversity Index by Hospital") +
                scale_y_continuous(limits = c(3.5,max(mge_shannon_div$shannon)), breaks = seq(3.5,max(mge_shannon_div$shannon), by = 0.5))

# Add ID as labels
p1 <- p1 + geom_text(aes(label = paste(rownames(mge_shannon_div), mge_shannon_div$shannon)), hjust = -0.1, vjust = 0.5, size = 4)

p1  # Display the plot

# Save plot
ggsave(filename = "plot7_mge_shannon.tiff",
       width = 16, height = 13, dpi = 300, units = "in", device='tiff', scale = 1)
```

#### Simpson for MGEs

```{r, warning=FALSE, error=FALSE, message=FALSE, results='hide', fig.show='hide'}
mge_simpson_div <- data.frame("simpson" = diversity(otu_table(MGE_PHY_counts_HWW), index = "simpson", MARGIN = 2))
all(rownames(mge_simpson_div) == colnames(MGE_PHY_counts_HWW@otu_table))
diversity_observed <- apply(MGE_PHY_counts_HWW@otu_table > 0, 2, sum)
mge_simpson_div$observed <- diversity_observed
mge_simpson_div$hospital <- as.factor(MGE_PHY_counts_HWW@sam_data$hospital)

# Get mean and standard deviation
mge_simpson_eda <- do.call(data.frame, aggregate(simpson ~ hospital, mge_simpson_div, function(x) c(mean = mean(x), sd = sd(x))))

# Test the significance
kruskal.test(simpson ~ hospital, data = mge_simpson_div)

# There are no sig. differences

# Test multiple pairwise
mge_simpson_stat.test <- pairwise.wilcox.test(mge_simpson_div$simpson, mge_simpson_div$hospital,
                 p.adjust.method = "BH")
mge_simpson_stat.test

# Plot
p <- ggplot(mge_simpson_div, aes(x=hospital, y=simpson, fill = hospital)) + 
  geom_boxplot()

# Add ID as label to the plot
p1 <- p + geom_dotplot(binaxis='y', stackdir='center', dotsize = 0.4) + 
  scale_fill_manual(values=c("#CD3857", "#61A6E3", "#CC7722")) +
  theme_classic() + theme(text = element_text(family = "Palatino"),
                          axis.title.y = element_text(size = 35),
                          axis.title.x = element_blank(),
                          axis.text.y = element_text(size = 20),
                          axis.text.x = element_text(size = 16, angle = 30, hjust = 1, face = "bold"),
                          legend.position = "none",
                          plot.title = element_text(face = "bold"),
                          plot.margin = margin(0.25, 0.25, 0.25, 0.25, "cm"))

# Adjust the scale to be more detailed
p1 <- p1 + labs(title = "Diversity Metrics for MGEs Across Hospitals", 
                subtitle = "Simpson Diversity Index by Hospital") +
                scale_y_continuous(limits = c(0.92,max(mge_simpson_div$simpson)), breaks = seq(0.92,max(mge_simpson_div$simpson), by = 0.02))

# Add ID as labels
p1 <- p1 + geom_text(aes(label = paste(rownames(mge_simpson_div), mge_simpson_div$simpson)), hjust = -0.1, vjust = 0.5, size = 4)

p1  # Display the plot

# Save plot
ggsave(filename = "plot7_mge_simpson.tiff",
       width = 16, height = 13, dpi = 300, units = "in", device='tiff', scale = 1)
```

#### Inverse Simpson for MGEs

```{r, warning=FALSE, error=FALSE, message=FALSE, results='hide', fig.show='hide'}
mge_invsimp_div <- data.frame("invsimpson" = diversity(otu_table(MGE_PHY_counts_HWW), index = "inv", MARGIN = 2))
all(rownames(mge_invsimp_div) == colnames(MGE_PHY_counts_HWW@otu_table))
diversity_observed <- apply(MGE_PHY_counts_HWW@otu_table > 0, 2, sum)
mge_invsimp_div$observed <- diversity_observed
mge_invsimp_div$hospital <- as.factor(MGE_PHY_counts_HWW@sam_data$hospital)

# Get mean and standard deviation
mge_invsimp_eda <- do.call(data.frame, aggregate(invsimpson ~ hospital, mge_invsimp_div, function(x) c(mean = mean(x), sd = sd(x))))

# Test the significance
kruskal.test(invsimpson ~ hospital, data = mge_invsimp_div)

# There are no sig. differences

# Test multiple pairwise
mge_invsimp_stat.test <- pairwise.wilcox.test(mge_invsimp_div$invsimpson, mge_invsimp_div$hospital,
                 p.adjust.method = "BH")
mge_invsimp_stat.test

# Plot
p <- ggplot(mge_invsimp_div, aes(x=hospital, y=invsimpson, fill = hospital)) + 
  geom_boxplot()

# Add ID as label to the plot
p1 <- p + geom_dotplot(binaxis='y', stackdir='center', dotsize = 0.4) + 
  scale_fill_manual(values=c("#CD3857", "#61A6E3", "#CC7722")) +
  theme_classic() + theme(text = element_text(family = "Palatino"),
                          axis.title.y = element_text(size = 35),
                          axis.title.x = element_blank(),
                          axis.text.y = element_text(size = 20),
                          axis.text.x = element_text(size = 16, angle = 30, hjust = 1, face = "bold"),
                          legend.position = "none",
                          plot.title = element_text(face = "bold"),
                          plot.margin = margin(0.25, 0.25, 0.25, 0.25, "cm"))

# Adjust the scale to be more detailed
p1 <- p1 + labs(title = "Diversity Metrics for MGEs Across Hospitals", 
                subtitle = "Inv Simpson Diversity Index by Hospital") +
                scale_y_continuous(limits = c(15,max(mge_invsimp_div$invsimpson)), breaks = seq(15,max(mge_invsimp_div$invsimpson), by = 10))

# Add ID as labels
p1 <- p1 + geom_text(aes(label = paste(rownames(mge_invsimp_div), mge_invsimp_div$invsimpson)), hjust = -0.1, vjust = 0.5, size = 4)

p1  # Display the plot

# Save plot
ggsave(filename = "plot7_mge_invsimpson.tiff",
       width = 16, height = 13, dpi = 300, units = "in", device='tiff', scale = 1)
```

#### Richness for MGEs

```{r, warning=FALSE, error=FALSE, message=FALSE, results='hide', fig.show='hide'}
mge_rich <- data.frame("richness" = specnumber(otu_table(MGE_PHY_counts_HWW), MARGIN=2))
all(rownames(mge_rich) == colnames(MGE_PHY_counts_HWW@otu_table))
mge_rich$hospital <- as.factor(MGE_PHY_counts_HWW@sam_data$hospital)

# Get mean and standard deviation
mge_rich_eda <- do.call(data.frame, aggregate(richness ~ hospital, mge_rich, function(x) c(mean = mean(x), sd = sd(x))))

# Test the significance
kruskal.test(richness ~ hospital, data = mge_rich)

# There are no sig. differences

# Test multiple pairwise
mge_rich_stat.test <- pairwise.wilcox.test(mge_rich$richness, mge_rich$hospital,
                 p.adjust.method = "BH")
mge_rich_stat.test

pvalues <- data.frame(group1 = c("SLMC","UST","CGH"), group2 = c("CGH","SLMC","UST"), p.adj = c("**","*","n.s."))

# Plot
p <- ggboxplot(mge_rich, x = "hospital", y = "richness", fill = "hospital", palette = c("#CD3857", "#61A6E3", "#CC7722")) + 
  stat_pvalue_manual(pvalues, size = 6, y.position = 900, step.increase = 0.05, label = "p.adj")

# Add ID as label to the plot
p1 <- p + geom_dotplot(binaxis='y', stackdir='center', dotsize = 0.4, aes(fill = hospital)) + 
  scale_fill_manual(values=c("#CD3857", "#61A6E3", "#CC7722")) +
  theme_classic() + theme(text = element_text(family = "Palatino"),
                          axis.title.y = element_text(size = 35),
                          axis.title.x = element_blank(),
                          axis.text.y = element_text(size = 20),
                          axis.text.x = element_text(size = 16, angle = 30, hjust = 1, face = "bold"),
                          legend.position = "none",
                          plot.title = element_text(face = "bold"),
                          plot.margin = margin(0.25, 0.25, 0.25, 0.25, "cm"))

# Adjust the scale to be more detailed
p1 <- p1 + labs(title = "Richness for MGEs Across Hospitals", 
                subtitle = "Richness by Hospital") +
                scale_y_continuous(limits = c(400,950), breaks = seq(400,950, by = 50))

# Add ID as labels
p1 <- p1 + geom_text(aes(label = paste(rownames(mge_rich), mge_rich$richness)), hjust = -0.1, vjust = 0.5, size = 4)

p1  # Display the plot

# Save plot
ggsave(filename = "plot7_mge_richness.tiff",
       width = 16, height = 13, dpi = 300, units = "in", device='tiff', scale = 1)
```

#### Evenness for MGEs

```{r, warning=FALSE, error=FALSE, message=FALSE, results='hide', fig.show='hide'}
mge_even <- data.frame("evenness" = mge_shannon_div$shannon / log(mge_rich$richness))
rownames(mge_even) <- colnames(MGE_PHY_counts_HWW@otu_table)
all(rownames(mge_even) == colnames(MGE_PHY_counts_HWW@otu_table))
mge_even$hospital <- as.factor(MGE_PHY_counts_HWW@sam_data$hospital)

# Get mean and standard deviation
mge_even_eda <- do.call(data.frame, aggregate(evenness ~ hospital, mge_even, function(x) c(mean = mean(x), sd = sd(x))))

# Test the significance
kruskal.test(evenness ~ hospital, data = mge_even)

# There are no sig. differences

# Test multiple pairwise
mge_even_stat.test <- pairwise.wilcox.test(mge_even$evenness, mge_even$hospital,
                 p.adjust.method = "BH")
mge_even_stat.test

# Plot
p <- ggplot(mge_even, aes(x=hospital, y=evenness, fill = hospital)) + 
  geom_boxplot()

# Add ID as label to the plot
p1 <- p + geom_dotplot(binaxis='y', stackdir='center', dotsize = 0.4) + 
  scale_fill_manual(values=c("#CD3857", "#61A6E3", "#CC7722")) +
  theme_classic() + theme(text = element_text(family = "Palatino"),
                          axis.title.y = element_text(size = 35),
                          axis.title.x = element_blank(),
                          axis.text.y = element_text(size = 20),
                          axis.text.x = element_text(size = 16, angle = 30, hjust = 1, face = "bold"),
                          legend.position = "none",
                          plot.title = element_text(face = "bold"),
                          plot.margin = margin(0.25, 0.25, 0.25, 0.25, "cm"))

# Adjust the scale to be more detailed
p1 <- p1 + labs(title = "Evenness for MGEs Across Hospitals", 
                subtitle = "Pielou's Evenness by Hospital") +
                scale_y_continuous(limits = c(0.65,max(mge_even$evenness)), breaks = seq(0.65,max(mge_even$evenness), by = 0.05))

# Add ID as labels
p1 <- p1 + geom_text(aes(label = paste(rownames(mge_even), mge_even$evenness)), hjust = -0.1, vjust = 0.5, size = 4)

p1  # Display the plot

# Save plot
ggsave(filename = "plot7_mge_evenness.tiff",
       width = 16, height = 13, dpi = 300, units = "in", device='tiff', scale = 1)
```

### Relative abundances & Counts of functional genes

#### Relative abundances of ARGs

```{r, warning=FALSE, error=FALSE, message=FALSE, results='hide', fig.show='hide'}
resfinder_PHY_rel_HWW <- subset_taxa(resfinder_PHY_rel_HWW, Gene != "NA")

p <- resfinder_PHY_rel_HWW %>%
  comp_barplot(tax_level = "Gene",
               n_taxa = 40,
               bar_outline_colour = NA,
               facet_by  = "hospital",
               bar_width = 0.7,
               other_name = "Others") +
  coord_flip()  +
  labs(x = NULL, y = NULL) + theme_light() + ggtitle("Relative Abundances of ARGs (Top 40)") +
  theme(text = element_text(family = "Palatino"),
        plot.title = element_text(size = 30, face = "bold"),
        axis.text.y = element_text(size = 10),
        axis.text.x = element_text(size = 20),
        strip.text.y = element_text(size = 20, color = "black", face = "bold", angle = 0, hjust = 0),
        strip.background = element_rect(fill = "white", color = "white"),
        panel.spacing.y = unit(1, "lines"),
        legend.text = element_text(size = 10, face = "italic"),
        legend.title = element_blank(),
        legend.position = "bottom", legend.direction = "horizontal",
        legend.spacing = unit(0.25, 'cm'),
        legend.key.size = unit(0.85, "cm")) + guides(fill = guide_legend(byrow = TRUE))

p

# Save plot
ggsave(filename = "plot8_arg_rel_gene.tiff",
       width = 16, height = 8, dpi = 300, units = "in", device='tiff', scale = 1)
```

#### Relative abundances of ARGs by drug class

```{r, warning=FALSE, error=FALSE, message=FALSE, results='hide', fig.show='hide'}
resfinder_PHY_rel_HWW <- subset_taxa(resfinder_PHY_rel_HWW, Class != "NA")

p <- resfinder_PHY_rel_HWW %>%
  comp_barplot(tax_level = "Class",
               n_taxa = 40,
               bar_outline_colour = NA,
               facet_by  = "hospital",
               bar_width = 0.7,
               other_name = "Others") +
  coord_flip()  +
  labs(x = NULL, y = NULL) + theme_light() + ggtitle("Relative Abundances of ARGs by Drug Class") +
  theme(text = element_text(family = "Palatino"),
        plot.title = element_text(size = 30, face = "bold"),
        axis.text.y = element_text(size = 10),
        axis.text.x = element_text(size = 20),
        strip.text.y = element_text(size = 20, color = "black", face = "bold", angle = 0, hjust = 0),
        strip.background = element_rect(fill = "white", color = "white"),
        panel.spacing.y = unit(1, "lines"),
        legend.text = element_text(size = 10, face = "italic"),
        legend.title = element_blank(),
        legend.position = "bottom", legend.direction = "horizontal",
        legend.spacing = unit(0.25, 'cm'),
        legend.key.size = unit(0.85, "cm")) + guides(fill = guide_legend(byrow = TRUE))

p

# Save plot
ggsave(filename = "plot8_arg_rel_class.tiff",
       width = 16, height = 8, dpi = 300, units = "in", device='tiff', scale = 1)
```

#### Count of ARGs

```{r, warning=FALSE, error=FALSE, message=FALSE, results='hide', fig.show='hide'}
resfinder_PHY_counts_HWW <- subset_taxa(resfinder_PHY_counts_HWW, Gene != "NA")

p <- resfinder_PHY_counts_HWW %>%
  comp_barplot(tax_level = "Gene",
               n_taxa = 40,
               tax_transform_for_plot = "identity",
               sample_order = sort(rownames(sam_data(resfinder_PHY_counts_HWW))),
               bar_outline_colour = NA,
               bar_width = 0.7,
               other_name = "Others") +
  labs(x = NULL, y = NULL) + theme_light() + ggtitle("Counts of ARGs (Top 40)") +
  theme(text = element_text(family = "Palatino"),
        plot.title = element_text(size = 30, face = "bold"),
        axis.text.y = element_text(size = 10),
        axis.text.x = element_text(size = 20, angle = 45, hjust = 1),
        strip.text.y = element_text(size = 20, color = "black", face = "bold", angle = 0, hjust = 0),
        strip.background = element_rect(fill = "white", color = "white"),
        panel.spacing.y = unit(1, "lines"),
        legend.text = element_text(size = 10, face = "italic"),
        legend.title = element_blank(),
        legend.position = "right",
        legend.direction = "horizontal",
        legend.spacing = unit(0.25, 'cm'),
        legend.key.size = unit(0.85, "cm")) + guides(fill = guide_legend(byrow = TRUE))

p

# Save plot
ggsave(filename = "plot9_arg_count_gene.tiff",
       width = 24, height = 8, dpi = 300, units = "in", device='tiff', scale = 1)
```

#### Count of ARGs

```{r, warning=FALSE, error=FALSE, message=FALSE, results='hide', fig.show='hide'}
resfinder_PHY_counts_HWW <- subset_taxa(resfinder_PHY_counts_HWW, Class != "NA")

p <- resfinder_PHY_counts_HWW %>%
  comp_barplot(tax_level = "Class",
               n_taxa = 40,
               tax_transform_for_plot = "identity",
               sample_order = sort(rownames(sam_data(resfinder_PHY_counts_HWW))),
               bar_outline_colour = NA,
               bar_width = 0.7,
               other_name = "Others") +
  labs(x = NULL, y = NULL) + theme_light() + ggtitle("Counts of ARGs by Drug Class") +
  theme(text = element_text(family = "Palatino"),
        plot.title = element_text(size = 30, face = "bold"),
        axis.text.y = element_text(size = 10),
        axis.text.x = element_text(size = 20, angle = 45, hjust = 1),
        strip.text.y = element_text(size = 20, color = "black", face = "bold", angle = 0, hjust = 0),
        strip.background = element_rect(fill = "white", color = "white"),
        panel.spacing.y = unit(1, "lines"),
        legend.text = element_text(size = 10, face = "italic"),
        legend.title = element_blank(),
        legend.position = "right",
        legend.direction = "horizontal",
        legend.spacing = unit(0.25, 'cm'),
        legend.key.size = unit(0.85, "cm")) + guides(fill = guide_legend(byrow = TRUE))

p

# Save plot
ggsave(filename = "plot9_arg_count_class.tiff",
       width = 24, height = 8, dpi = 300, units = "in", device='tiff', scale = 1)
```

```{r, warning=FALSE, error=FALSE, message=FALSE, results='hide', fig.show='hide'}
# Gather data

#PKY: Is this the correct way? We got 105% for CGH2...
colSums(otu_table(resfinder_PHY_rel_HWW))

df <- data.frame(
  RA = sample_sums(resfinder_PHY_rel_HWW),
  hospital = as.factor(sample_data(resfinder_PHY_rel_HWW)$hospital)
)
M <- lm(RA ~ hospital, data = df)
summary(M)
shapiro.test(df$RA)
# Data is not normally distributed.
plot(M)

# Heterosedasticity?
par(mfrow=c(2,2))
#plot(M)

lmtest::bptest(M)  # Breusch-Pagan test
car::ncvTest(M)  # Breusch-Pagan test
# There is no heteroscedasticity.
```

```{r}
# Test
kruskal.test(RA ~ hospital, data = df)
```

```{r}
# Test multiple pairwise
stat.test <- pairwise.wilcox.test(df$RA, df$hospital,
                 p.adjust.method = "BH")
stat.test

pvalues <- tibble::tribble(
  ~group1, ~group2,   ~p.adj,
    "SLMC",     "CGH", "***",
    "SLMC",     "UST", "**",
    "CGH",     "UST", "**"
  )
pvalues

# +
#   stat_pvalue_manual(pvalues, hide.ns = TRUE, size = 8, y.position = 2.5, step.increase = 0.025, label = "p.adj")

# Save plot
p <- ggboxplot(df, x = "hospital", y = "RA", color = "hospital", palette = c("#1C4731", "#910C28", "#1E88E5"), width = 0.3)  +
  theme_bw() +
  theme(axis.text.x = element_text(size = 16, face = "bold"),
        axis.title.x = element_blank(),
        axis.text.y = element_text(size = 16),
        axis.title.y = element_text(size = 20),
        legend.position = "none",
        plot.subtitle = element_text(size = 24, hjust = 0.5),
        plot.tag = element_text(size = 24, family = "Palatino", face = "bold"),
        text = element_text(family = "Palatino"),
        aspect.ratio = 1) +
  labs(y = "ARGs/16S rRNA", x = "") +
  guides(color = "none", alpha = "none") +
  geom_jitter(data = df, aes(x = hospital, y = RA, color = hospital), size = 3, alpha = 0.7, width = 0.1) +
  geom_text(data = df, aes(x = hospital, y = RA, label = rownames(arg_shannon_div)), size = 3, vjust = -0.5)  +
  labs(title = "Comparison of ARGs/16S rRNA Levels Across Hospitals using Wilcoxx")  # Add title

ggsave(filename = "wilcox.tiff",
       width = 16, height = 13, dpi = 300, units = "in", device='tiff', scale = 1)
```

#### Relative abundances of MGEs

```{r, warning=FALSE, error=FALSE, message=FALSE, results='hide', fig.show='hide'}
MGE_rel_PHY <- subset_taxa(MGE_rel_PHY, Gene != "NA")

p <- MGE_rel_PHY %>%
  comp_barplot(tax_level = "Gene",
               n_taxa = 40,
               bar_outline_colour = NA,
               facet_by  = "hospital",
               bar_width = 0.7,
               other_name = "Others") +
  coord_flip()  +
  labs(x = NULL, y = NULL) + theme_light() + ggtitle("Relative Abundances of MGEs (Top 40)") +
  theme(text = element_text(family = "Palatino"),
        plot.title = element_text(size = 30, face = "bold"),
        axis.text.y = element_text(size = 10),
        axis.text.x = element_text(size = 20),
        strip.text.y = element_text(size = 20, color = "black", face = "bold", angle = 0, hjust = 0),
        strip.background = element_rect(fill = "white", color = "white"),
        panel.spacing.y = unit(1, "lines"),
        legend.text = element_text(size = 10, face = "italic"),
        legend.title = element_blank(),
        legend.position = "bottom", legend.direction = "horizontal",
        legend.spacing = unit(0.25, 'cm'),
        legend.key.size = unit(0.85, "cm")) + guides(fill = guide_legend(byrow = TRUE))

p

# Save plot
ggsave(filename = "plot10_mge_rel_gene.tiff",
       width = 16, height = 8, dpi = 300, units = "in", device='tiff', scale = 1)
```

#### Relative abundances of Taxa (metaphlan)

##### Species level

```{r, warning=FALSE, error=FALSE, message=FALSE, results='hide', fig.show='hide'}
metaphlan_PHY_rel_species <- subset_taxa(metaphlan_PHY_rel_species, Species != "NA")

metaphlan_PHY_rel_species %>%
 tax_fix(
  min_length = 4,
  unknowns = c("NA"),
  sep = " ", anon_unique = TRUE,
  suffix_rank = "classified"
 )

p <- metaphlan_PHY_rel_species %>%
  comp_barplot(tax_level = "Species",
               n_taxa = 40,
               bar_outline_colour = NA,
               facet_by  = "hospital",
               bar_width = 0.7,
               other_name = "Others") +
  coord_flip()  +
  labs(x = NULL, y = NULL) + theme_light() + ggtitle("Relative Abundances of Top 40 Species (metaphlan)") +
  theme(text = element_text(family = "Palatino"),
        plot.title = element_text(size = 30, face = "bold"),
        axis.text.y = element_text(size = 10),
        axis.text.x = element_text(size = 20),
        strip.text.y = element_text(size = 20, color = "black", face = "bold", angle = 0, hjust = 0),
        strip.background = element_rect(fill = "white", color = "white"),
        panel.spacing.y = unit(1, "lines"),
        legend.text = element_text(size = 10, face = "italic"),
        legend.title = element_blank(),
        legend.position = "bottom", legend.direction = "horizontal",
        legend.spacing = unit(0.25, 'cm'),
        legend.key.size = unit(0.85, "cm")) + guides(fill = guide_legend(byrow = TRUE))

p

# Save plot
ggsave(filename = "plot12_tax_rel_species.tiff",
       width = 16, height = 8, dpi = 300, units = "in", device='tiff', scale = 1)
```

##### Genus level

```{r, warning=FALSE, error=FALSE, message=FALSE, results='hide', fig.show='hide'}
metaphlan_PHY_rel_genus <- subset_taxa(metaphlan_PHY_rel_genus, Genus != "Coprobacillaceae Family")
metaphlan_PHY_rel_genus <- subset_taxa(metaphlan_PHY_rel_genus, Genus != "Catenibacterium")
metaphlan_PHY_rel_genus <- subset_taxa(metaphlan_PHY_rel_genus, Genus != "NA")

metaphlan_PHY_rel_genus %>%
 tax_fix(
  min_length = 4,
  unknowns = c("NA"),
  sep = " ", anon_unique = TRUE,
  suffix_rank = "classified"
 )

p <- metaphlan_PHY_rel_genus %>%
  comp_barplot(tax_level = "Genus",
               n_taxa = 40,
               bar_outline_colour = NA,
               facet_by  = "hospital",
               bar_width = 0.7,
               other_name = "Others") +
  coord_flip()  +
  labs(x = NULL, y = NULL) + theme_light() + ggtitle("Relative Abundances of Top 40 Genera (metaphlan)") +
  theme(text = element_text(family = "Palatino"),
        plot.title = element_text(size = 30, face = "bold"),
        axis.text.y = element_text(size = 10),
        axis.text.x = element_text(size = 20),
        strip.text.y = element_text(size = 20, color = "black", face = "bold", angle = 0, hjust = 0),
        strip.background = element_rect(fill = "white", color = "white"),
        panel.spacing.y = unit(1, "lines"),
        legend.text = element_text(size = 12, face = "italic"),
        legend.title = element_blank(),
        legend.position = "bottom", legend.direction = "horizontal",
        legend.spacing = unit(0.25, 'cm'),
        legend.key.size = unit(0.85, "cm")) + guides(fill = guide_legend(byrow = TRUE))

p

# Save plot
ggsave(filename = "plot12_tax_rel_genus.tiff",
       width = 16, height = 8, dpi = 300, units = "in", device='tiff', scale = 1)
```

##### Family level

```{r, warning=FALSE, error=FALSE, message=FALSE, results='hide', fig.show='hide'}
metaphlan_PHY_rel_family <- subset_taxa(metaphlan_PHY_rel_family, Family != "Coprobacillaceae")
metaphlan_PHY_rel_family <- subset_taxa(metaphlan_PHY_rel_family, Family != "NA")

metaphlan_PHY_rel_family %>%
 tax_fix(
  min_length = 4,
  unknowns = c("NA"),
  sep = " ", anon_unique = TRUE,
  suffix_rank = "classified"
 )

p <- metaphlan_PHY_rel_family %>%
  comp_barplot(tax_level = "Family",
               n_taxa = 40,
               bar_outline_colour = NA,
               facet_by  = "hospital",
               bar_width = 0.7,
               other_name = "Others") +
  coord_flip()  +
  labs(x = NULL, y = NULL) + theme_light() + ggtitle("Relative Abundances of Top 40 Families (metaphlan)") +
  theme(text = element_text(family = "Palatino"),
        plot.title = element_text(size = 30, face = "bold"),
        axis.text.y = element_text(size = 10),
        axis.text.x = element_text(size = 20),
        strip.text.y = element_text(size = 20, color = "black", face = "bold", angle = 0, hjust = 0),
        strip.background = element_rect(fill = "white", color = "white"),
        panel.spacing.y = unit(1, "lines"),
        legend.text = element_text(size = 12, face = "italic"),
        legend.title = element_blank(),
        legend.position = "bottom", legend.direction = "horizontal",
        legend.spacing = unit(0.25, 'cm'),
        legend.key.size = unit(0.85, "cm")) + guides(fill = guide_legend(byrow = TRUE))

p

# Save plot
ggsave(filename = "plot12_tax_rel_family.tiff",
       width = 16, height = 8, dpi = 300, units = "in", device='tiff', scale = 1)
```

##### Class level

```{r, warning=FALSE, error=FALSE, message=FALSE, results='hide', fig.show='hide'}
metaphlan_PHY_rel_class <- subset_taxa(metaphlan_PHY_rel_class, Class != "Clostridia")
metaphlan_PHY_rel_class <- subset_taxa(metaphlan_PHY_rel_class, Class != "Erysipelotrichia")
metaphlan_PHY_rel_class <- subset_taxa(metaphlan_PHY_rel_class, Class != "NA")

metaphlan_PHY_rel_class %>%
 tax_fix(
  min_length = 4,
  unknowns = c("NA"),
  sep = " ", anon_unique = TRUE,
  suffix_rank = "classified"
 )

p <- metaphlan_PHY_rel_class %>%
  comp_barplot(tax_level = "Class",
               n_taxa = 40,
               bar_outline_colour = NA,
               facet_by  = "hospital",
               bar_width = 0.7,
               other_name = "Others") +
  coord_flip()  +
  labs(x = NULL, y = NULL) + theme_light() + ggtitle("Relative Abundances of Top 40 Classes (metaphlan)") +
  theme(text = element_text(family = "Palatino"),
        plot.title = element_text(size = 30, face = "bold"),
        axis.text.y = element_text(size = 10),
        axis.text.x = element_text(size = 20),
        strip.text.y = element_text(size = 20, color = "black", face = "bold", angle = 0, hjust = 0),
        strip.background = element_rect(fill = "white", color = "white"),
        panel.spacing.y = unit(1, "lines"),
        legend.text = element_text(size = 12, face = "italic"),
        legend.title = element_blank(),
        legend.position = "bottom", legend.direction = "horizontal",
        legend.spacing = unit(0.25, 'cm'),
        legend.key.size = unit(0.85, "cm")) + guides(fill = guide_legend(byrow = TRUE))

p

# Save plot
ggsave(filename = "plot12_tax_rel_class.tiff",
       width = 16, height = 8, dpi = 300, units = "in", device='tiff', scale = 1)
```

##### Phylum level

```{r, warning=FALSE, error=FALSE, message=FALSE, results='hide', fig.show='hide'}
metaphlan_PHY_rel_phylum <- subset_taxa(metaphlan_PHY_rel_phylum, Phylum != "NA")

metaphlan_PHY_rel_phylum %>%
 tax_fix(
  min_length = 4,
  unknowns = c("NA"),
  sep = " ", anon_unique = TRUE,
  suffix_rank = "classified"
 )

p <- metaphlan_PHY_rel_phylum %>%
  comp_barplot(tax_level = "Phylum",
               n_taxa = 40,
               bar_outline_colour = NA,
               facet_by  = "hospital",
               bar_width = 0.7,
               other_name = "Others") +
  coord_flip()  +
  labs(x = NULL, y = NULL) + theme_light() + ggtitle("Relative Abundances of Phyla (metaphlan)") +
  theme(text = element_text(family = "Palatino"),
        plot.title = element_text(size = 30, face = "bold"),
        axis.text.y = element_text(size = 10),
        axis.text.x = element_text(size = 20),
        strip.text.y = element_text(size = 20, color = "black", face = "bold", angle = 0, hjust = 0),
        strip.background = element_rect(fill = "white", color = "white"),
        panel.spacing.y = unit(1, "lines"),
        legend.text = element_text(size = 12, face = "italic"),
        legend.title = element_blank(),
        legend.position = "bottom", legend.direction = "horizontal",
        legend.spacing = unit(0.25, 'cm'),
        legend.key.size = unit(0.85, "cm")) + guides(fill = guide_legend(byrow = TRUE))

p

# Save plot
ggsave(filename = "plot12_tax_rel_phylum.tiff",
       width = 16, height = 8, dpi = 300, units = "in", device='tiff', scale = 1)
```

Add a new chunk by clicking the *Insert Chunk* button on the toolbar or by pressing *Ctrl+Alt+I*.

When you save the notebook, an HTML file containing the code and output will be saved alongside it (click the *Preview* button or press *Ctrl+Shift+K* to preview the HTML file).

The preview shows you a rendered HTML copy of the contents of the editor. Consequently, unlike *Knit*, *Preview* does not run any R code chunks. Instead, the output of the chunk when it was last run in the editor is displayed.

```{r}
p<-plot_bar(resfinder_PHY_rel_HWW, fill = "Class", x = "hospital")

ggsave(filename = "bar_Resfinder.tiff",
       width = 16, height = 13, dpi = 300, units = "in", device='tiff', scale = 1)
```

```{r}
otu_table_data <- otu_table(resfinder_PHY_rel_HWW)
print(head(otu_table_data))
```
